src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1295 @@
+      state_label_map_[tok] = id;
+      // The final weight has been worked out in the previous for loop and
+      // store in the states
+      // Here, we create a specific final state, and move the final costs to
+      // the cost of an arc connecting to the final state
+      KALDI_ASSERT(ofst->Final(cur_state) != Weight::Zero());
+      Weight final_weight = ofst->Final(cur_state);
+      // Use cost_offsets to guide DeterminizeLatticePruned()
+      // later in GetLattice()
+      // For now, we use extra_cost from the decoding stage , which has some
+      // "future information", as
+      // the final weights of this chunk
+      BaseFloat cost_offset = tok->extra_cost;
+      // We record these cost_offset, and after we appending two chunks
+      // we will cancel them out
+      state_label_final_cost_[id] = cost_offset;
+      Arc arc(0, id, Times(final_weight, Weight(0, cost_offset)), end_state);
+      ofst->AddArc(cur_state, arc);
+      ofst->SetFinal(cur_state, Weight::Zero());
+    }
+  }
+  return (ofst->NumStates() > 0);
+}
+
+template <typename FST, typename Token>
+void LatticeIncrementalDecoderTpl<FST, Token>::AppendLatticeChunks(
+    CompactLattice clat, bool not_first_chunk, int32 last_frame_of_chunk,
+    CompactLattice *olat) {
+  using namespace fst;
+  // step 3.1: Appending new chunk to the old one
+  int32 state_offset = olat->NumStates();
+  if (not_first_chunk)
+    state_offset--; // since we do not append initial state in the first chunk
+
+  // A map between state label and the arc position (index)
+  // the incoming states of these arcs are initial states of the chunk
+  // and the olabel of these arcs are the key of this map (state label)
+  // The arc position are obtained from ArcIterator corresponding to the state
+  unordered_map<int32, size_t> initial_arc_map;
+  initial_arc_map.reserve(std::min((int32)1e5, config_.max_active));
+  for (StateIterator<CompactLattice> siter(clat); !siter.Done(); siter.Next()) {
+    auto s = siter.Value();
+    StateId state_appended = -1;
+    // We do not copy initial state, which exists except the first chunk
+    if (!not_first_chunk || s != 0) {
+      state_appended = s + state_offset;
+      KALDI_ASSERT(state_appended == olat->AddState());
+      olat->SetFinal(state_appended, clat.Final(s));
+    }
+
+    for (ArcIterator<CompactLattice> aiter(clat, s); !aiter.Done(); aiter.Next()) {
+      const auto &arc = aiter.Value();
+      // We do not copy initial arcs, which exists except the first chunk.
+      // These arcs will be taken care later in step 3.2
+      if (!not_first_chunk || s != 0) {
+        CompactLatticeArc arc_appended(arc);
+        arc_appended.nextstate += state_offset;
+        olat->AddArc(state_appended, arc_appended);
+      }
+      // Process state labels, which will be used in step 3.2
+      if (arc.olabel > config_.max_word_id) { // initial_arc
+        if (s == 0) { // record initial_arc in this chunk, we will use it right now
```

Comment:
@danpovey  I found that for the first chunk, there could be some final arcs starting from state 0, while for the last chunk, there could be some initial arcs ending in final state. Hence, I found that we cannot distinguish final and initial arcs by simply "if (s==0)" or "if (clat.Final(arc_appended.nextstate)!=CompactLatticeWeight::Zero()"
How about we make the state label of a initial arc = the corresponding state label of its final arc + a certain offset.
By this way we can explicitly distinguish final and initial arcs in clat here

Reply:
Hm.  Don't you know, in this part of the code, whether this chunk is initial and whether it is final (meaning: whether it is the first chunk, and whether it is the last chunk)?

You could use logic like:
```
  if (s == 0 && !is_initial_chunk) {
     // process initial arc (meaning: arc at the beginning of a chunk with a state label on it
  } else {
    // process final arc (meaning: arc at the end of the chunk with a state label on it).
  }
```   
that would avoid the hassle of messing around with states.

Reply:
Sure I can do something like this if you prefer